src/state.rs

```diff
@@ -72,6 +73,46 @@ pub struct ProxyState {
 
     #[serde(flatten)]
     pub policies: PolicyStore,
+
+    pub resolved_dns: ResolvedDnsStore,
+}
+
+/// A ResolvedDnsStore encapsulates all per-workload resolved DNS information for workloads in the mesh
+#[derive(serde::Serialize, Default, Debug)]
+pub struct ResolvedDnsStore {
+    // workload UID to resolved IP addresses
+    by_workload_uid: HashMap<String, HashSet<IpAddr>>,
+}
+
+impl ResolvedDnsStore {
+    pub fn load_balance_for_workload(&self, workload_uid: String) -> Result<IpAddr, Error> {
+        debug!(
+            "load_balance_for_workload({}) by_workload_uid len: {}",
+            &workload_uid,
+            self.by_workload_uid.len()
+        );
+        // TODO: add support for different async dns modes: e.g. logical DNS vs strict DNS.
+        // this is effectively envoy strict DNS; returning a pre-resolved IP address at random.
+        let Some(ipset) = self.by_workload_uid.get(&workload_uid) else {
+            return Err(Error::NoResolvedAddresses(workload_uid));
+        };
+        let ips_vec = ipset.iter().collect::<Vec<_>>();
+        let Some(ip) = ips_vec.choose(&mut rand::thread_rng()) else {
+            return Err(Error::EmptyResolvedAddresses(workload_uid));
+        };
+        Ok(**ip)
+    }
+
+    // TODO: ipv6 support
+    pub fn set_ips_for_workload(&mut self, workload_uid: String, ips: Vec<Ipv4Addr>) {
+        let set = HashSet::from_iter(ips.iter().map(|x| IpAddr::V4(*x)));
+        self.by_workload_uid.insert(workload_uid.to_owned(), set);
+        trace!(
+            "set_ips_for_workload({}), by_workload_uid len: {}",
```

Comment:
nit: Might be more appropriate to use things like https://docs.rs/tracing/latest/tracing/attr.instrument.html for this.

Basically we are abusing a string to trace a function call + add k/v attributes, both of which the library can handle natively/better.

NBD especially for trace, more a comment for awareness

Reply:
simplified and removed this code